hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/producers/BlockStreamProducerSingleThreaded.java

```diff
@@ -0,0 +1,248 @@
+            // BlockStreamWriterFactory instantiated by dagger will determine this.
+            writer.writeItem(serializedItem, runningHash);
+        } catch (final Exception e) {
+            // This **may** prove fatal. The node should be able to carry on, but then fail when it comes to
+            // actually producing a valid record stream file. We need to have some way of letting all nodesknow
+            // that this node has a problem, so we can make sure at least a minimal threshold of nodes is
+            // successfully producing a blockchain.
+            logger.error("Error writing block item to block stream writer for block {}", currentBlockNumber, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void initFromLastBlock(@NonNull RunningHashes runningHashes, long lastBlockNumber) {
+        if (this.runningHash != null) {
+            throw new IllegalStateException("initFromLastBlock() must only be called once");
+        }
+
+        this.runningHash = runningHashes.runningHash();
+        this.runningHashNMinus1 = runningHashes.nMinus1RunningHash();
+        this.runningHashNMinus2 = runningHashes.nMinus2RunningHash();
+        this.runningHashNMinus3 = runningHashes.nMinus3RunningHash();
+        this.currentBlockNumber = lastBlockNumber;
+    }
+
+    @NonNull
+    @Override
+    public Bytes getRunningHash() {
+        assert runningHash != null : "initFromLastBlock() must be called before getRunningHash()";
+        return runningHash;
+    }
+
+    @Nullable
+    @Override
+    public Bytes getNMinus3RunningHash() {
+        return runningHashNMinus3;
+    }
+
+    @Override
+    public void beginBlock() {
+        this.currentBlockNumber++;
+
+        final var lastRunningHash = getRunningHashObject();
+
+        openWriter(this.currentBlockNumber, lastRunningHash);
+
+        // Write the block header to the block stream as the first item in the block.
+        writeBlockHeader(lastRunningHash);
+    }
+
+    @Override
+    public void endBlock() {
+        final var lastRunningHash = getRunningHashObject();
+        closeWriter(lastRunningHash, this.currentBlockNumber);
+    }
+
+    @Override
+    public void writeSystemTransaction(ConsensusTransaction platformTxn) {
+        final var serializedBlockItem = format.serializeSystemTransaction(platformTxn);
+        updateRunningHashes(serializedBlockItem);
+        writeSerializedBlockItem(serializedBlockItem);
+    }
+
+    @Override
+    public void close() throws Exception {
+        // FUTURE: close() should wait until the block is completed, which cannot happen until the block proof is
+        // produced, and that cannot happen until the signatures have been gossiped. Today, this will close in
+        // unpredictable ways, and will most likely result in the block being written incompletely without a block
+        // proof.
```

Comment:
I suspect we'll end up collecting signatures for the block proof in the block manager, and only call this `close()` when we have sufficient signatures. What do you think, though?